pkg/common/watchdog.go

```diff
@@ -0,0 +1,70 @@
+package common
+
+import (
+	"sync"
+	"time"
+)
+
+// Configuration for the watchdog.
+type WatchdogConfig struct {
+	// Timeout after which `OnTimeout` is called.
+	Timeout time.Duration
+	// A closure that is called once `Timeout` is reached.
+	OnTimeout func()
+}
+
+// We need to wrap the channel in a struct so that we can close it from the outside and
+// check by the sender if the channel is closed (there is no elegant way to do it in Go).
+type WatchdogChannel struct {
+	channel chan<- struct{}
+	mutex   sync.Mutex
+	closed  bool
+}
+
+// Close the channel unless already closed.
+func (c *WatchdogChannel) Close() {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if !c.closed {
+		close(c.channel)
+		c.closed = true
+	}
+}
+
+// Notify the watchdog that a packet has been received. Return `true` if the notification
+// has been sent, `false` if the channel is already closed.
+func (c *WatchdogChannel) Notify() bool {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if !c.closed {
+		c.channel <- struct{}{}
+		return true
+	}
+
+	return false
+}
+
+// Starts a watchdog that periodically (specified by the configuration) executes a `c.OnTimeout` closure if
+// no messages have been received on a channel for a `c.Timeout`.
+func (c *WatchdogConfig) Start() *WatchdogChannel {
```

Comment:
The usage of `WatchdogConfig` here feels odd to me, wouldn't a `NewWatchdogChannel(c WatchdogConfig) *WatchdogChannel {}` be better? Or is this the go way to do things?

Reply:
We did something similar in `HeartBeat`, so I just followed the same approach, but you're probably right, perhaps it's better to have an explicit `NewWatchdog()` function that accepts the config, let me rework that real quick.